### DISCUSSION TITLE
How to connect a database (via Mongoose) with middleware before running apollo-server-micro in an API route?

### DISCUSSION THREAD

**Discussion by EddyVinck** [vercel/next.js#13280] [Help]
## Description

I based my code on the [API routes with apollo server and client with auth example.](https://github.com/zeit/next.js/tree/canary/examples/api-routes-apollo-server-and-client-auth). There is still some code in my repo from the [`passport` example](https://github.com/zeit/next.js/tree/canary/examples/with-passport) like additional API routes besides `api/graphql.js` but those can be ignored.

I'm trying to connect to my database before my GraphQL API route is called, but I can't get it to work so my data isn't fetched during SSR. 

I'm still kind of new to API routes and I tried following a few examples that tried to do similar things online, but I couldn't get it to work. I have been at this all day and I'm not sure how to fix it.

## To Reproduce

Repository: https://github.com/EddyVinck/nextjs-graphql-boilerplate/tree/api-route-db-repro

Steps to reproduce the behavior, on the `api-route-db-repro` branch:

1. Make sure `mongod` is running locally
2. run `yarn dev`
3. go to `http://localhost:3000/`
4. See `"loading posts"` in the DOM (or view-source) 

## Expected behavior

The data should be fetched during server-side rendering, so the `"loading posts"` text should not appear when checking the source of the page.

## Screenshots & code

[My database middleware:](https://github.com/EddyVinck/nextjs-graphql-boilerplate/blob/api-route-db-repro/middlewares/database.js)
```javascript
export const database = (handler) => async (req, res, ...restArgs) => {
  if (mongoose.connection.readyState !== MONGOOSE_CONNECTED) {
    connection = await connect(dbUrl);

    if (IS_DEV && connection) {
      await createSampleDataIfDbEmpty();
    }
  }
  req.db = connection;
  return handler(req, res, ...restArgs);
};
```
[My GraphQL API route:](https://github.com/EddyVinck/nextjs-graphql-boilerplate/blob/api-route-db-repro/pages/api/graphql.js)
```javascript
// apollo-server-micro
const apolloServer = new ApolloServer({
  schema,
  context: async ({ req, res }) => {
    const ctx = { req, res, db: null, user: null };
    ctx.db = req.db;
    const user = await getUserFromReq(req);
    ctx.user = user;
    return ctx;
  },
});

export const config = {
  api: {
    bodyParser: false,
  },
};

// from micro-cors
const cors = Cors({
  allowMethods: ["GET", "POST", "OPTIONS"],
});

const apollo = apolloServer.createHandler({ path: "/api/graphql" });
const handler = database(cors(apollo));

export default handler;
```

For some reason it sometimes does have the database in the context on the first refresh after running `yarn dev` (see `"ctx.db"` log):

![image](https://user-images.githubusercontent.com/23434753/82737212-99591f80-9d2f-11ea-985c-1e2a1a920a4d.png)

## System information

- OS: Ubuntu 20.4
- Browser: Chrome
- Version of Next.js: "next": "9.4.1",
- Version of Node.js: 12.16.3

## Additional context

Like I said, I've been at this all day trying to figure this all out, so maybe I'm just not seeing something silly right now. 

Maybe `next-connect` can help here? I can't think straight anymore at this point today, so I don't really know.

**Comment by EddyVinck** [comment]
It seems multiple people are unsure how to do this:
- https://github.com/zeit/next.js/discussions/12294
- https://github.com/zeit/next.js/discussions/12229

I found [this reply](https://github.com/zeit/next.js/discussions/11024#discussioncomment-1330) but it doesn't apply here since my API is much different and it seems silly to duplicate the `apollo-server-micro` on every page instead of utilizing API routes. It's also worth noting that the `micro` version of this `apollo-server` [is not a requirement](https://github.com/zeit/next.js/discussions/11024#discussioncomment-1330), although most tutorials and examples do use `apollo-server-micro`. 

Someone suggested [on Twitter](https://twitter.com/diondre27/status/1264914445660229633) to connect in the `context` function, but I don't think that's a good idea if the connection can be cached as opposed to reinstantiated on every request (as is the case with the context function which runs on every request).

It would be great if this could work as an API route, but if it needs to be done with a custom server that's also fine by me. Some help with this would be very much appreciated.

**Comment by divslingerx** [comment]
I'm not sure If I did this the correct what but i just made my db connection into a middleware of sorts. 

#mongo/connect.js
```
import mongoose from "mongoose"

const connect = async () => {
   await mongoose
  .connect(process.env.MONGO_URI, { useNewUrlParser: true, useFindAndModify: false, useUnifiedTopology: true})
  .then(() => console.log(`Mongo running at ${process.env.MONGO_URI}`))
  .catch(err => console.log(err))
}

const connectDB = (handler) => async (req, res) => {
    if(mongoose.connections[0].readyState !== 1){
        await connect()
    }
    return handler(req,res)
}

const db = mongoose.connection
db.once('ready', () => console.log(`connected to mongo on ${process.env.MONGO_URI}`))

export default connectDB
```



export default connectDB

and then in the server I do 
`connectDB(apolloServer.createHandler({ path: "/api/graphql" }))`

**Comment by Vince-vegas** [comment]
my config file

/utils/connectDB
```
import mongoose from 'mongoose';

const connectDB = async () => {
  if (mongoose.connections[0].readyState) {
    console.log('already connected');
    return;
  }

  mongoose
    .connect(process.env.MONGODB_URI,
      {
        useNewUrlParser: true,
        useUnifiedTopology: true,
        useFindAndModify: false,
        useCreateIndex: true,
      }
    )
    .then(() => console.log('database started'))
    .catch((err) => console.log(err));

  //
};

export default connectDB;
```

/models/Collection
```
// Collection Schema
const Collection = mongoose.models.collection || mongoose.model('collection', CollectionSchema);

export default Collection;
```

/api
in API folder in nextjs call the function

```
import connectDB from '../utils/connectDB';
import Collection from '../models/Collection';

connectDB()

export default (req, res) => {
 // fetch data from database collection
 Collection.find()
}
```
